File: src/de/kasoki/trierbustimetracker/BusTimeActivity.java
package de.kasoki.trierbustimetracker;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.StrictMode;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ListView;
import android.widget.Toast;
import de.kasoki.swtrealtime.BusStop;
import de.kasoki.trierbustimetracker.adapter.BusTimeAdapter;
import de.kasoki.trierbustimetracker.tasks.ReloadTask;
import de.kasoki.trierbustimetracker.utils.Helper;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class BusTimeActivity extends Activity {

	private volatile ArrayList<HashMap<String, String>> listViewContent;

	private volatile ListView busStopListView;
	private volatile BusTimeAdapter listAdapter;

	private String busTimeCode;

	private volatile boolean reloadActive = false;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_bustimes);

		if(Helper.getCurrentAPILevel() >= 11) {
			this.addHomeAsUpButtonToActionBar();
		}
		
		// get selected bus code
		Intent intent = getIntent();
		busTimeCode = intent.getStringExtra("BUS_TIME_CODE");

		// init stuff
		busStopListView = (ListView) this.findViewById(R.id.busStopListView);
		listViewContent = new ArrayList<HashMap<String, String>>();

		listAdapter = new BusTimeAdapter(listViewContent, this);

		busStopListView.setAdapter(listAdapter);

		if(Helper.getCurrentAPILevel() >= 9) {
			this.setThreadPolicyToAllowAll();
		}

		// set title to bus stop
		this.setTitle(BusStop.getBusStopByStopCode(busTimeCode).getName());

		// get information
		reload();
	}

	@TargetApi(Build.VERSION_CODES.HONEYCOMB)
	private void addHomeAsUpButtonToActionBar() {
		getActionBar().setDisplayHomeAsUpEnabled(true);
	}

	@TargetApi(Build.VERSION_CODES.GINGERBREAD)
	private void setThreadPolicyToAllowAll() {
		// set thread policy to permit all
		StrictMode.ThreadPolicy policy = new StrictMode.ThreadPolicy.Builder()
				.permitAll().build();
		StrictMode.setThreadPolicy(policy);
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		// Inflate the menu; this adds items to the action bar if it is present.
		getMenuInflater().inflate(R.menu.bustimes_menu, menu);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
		// reload action pressed
		case R.id.action_reload:
			reload();
			return true;
		case android.R.id.home:
			finish();
			return true;
		default:
			return super.onOptionsItemSelected(item);
		}
	}

	// reload the bus times and set them to the busStopListView
	private synchronized void reload() {
		final String busTimeCode = this.busTimeCode;
		final BusTimeActivity activity = this;

		if (!this.reloadActive) {
			if (Helper.isNetworkAvailable(activity)) {
				new ReloadTask(activity, busTimeCode).execute(0);
			} else {
				// No connection
				Log.d("NETWORK", "NO NETWORK CONNECTION");

				String noNetworkConnectionText = activity.getResources()
						.getString(R.string.no_network_connection_text);

				Toast toast = Toast.makeText(activity.getApplicationContext(),
						noNetworkConnectionText, Toast.LENGTH_SHORT);
				toast.show();
			}
		}
	}

	public void setReloadActive(boolean bool) {
		this.reloadActive = bool;
	}

	public void notifyListViewDataSetChanged() {
		listAdapter.notifyDataSetChanged();
	}

	public void setListViewContent(List<HashMap<String, String>> content) {
		listViewContent.clear();

		for (HashMap<String, String> data : content) {
			listViewContent.add(data);
		}
	}

	public void onReloadTaskFinished(final List<HashMap<String, String>> content) {
		final BusTimeActivity activity = this;

		// run code in the main thread
		Handler mainHandler = new Handler(activity.getApplicationContext()
				.getMainLooper());

		Runnable r = new Runnable() {
			public void run() {
				activity.setListViewContent(content);

				activity.notifyListViewDataSetChanged();

			}
		};

		mainHandler.post(r);
	}

    public void onReloadTaskFailed() {
        this.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if(listViewContent.isEmpty()) {
                    HashMap<String, String> data = new HashMap<String, String>();

                    data.put("NUMBER", "");
                    data.put("DESTINATION", getString(R.string.connection_failed));
                    data.put("TIME", "");

                    listViewContent.add(data);

                    notifyListViewDataSetChanged();
                }
            }
        });
    }

    public void toastServerError(final int responseCode) {
        this.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Toast.makeText(BusTimeActivity.this, getString(R.string.server_response_error, responseCode), Toast.LENGTH_SHORT).show();
            }
        });
    }

	@Override
	public void onResume() {
		reload();

		super.onResume();
	}

}

File: src/main/scala/de/kasoki/trierbustimetracker/BusTimeActivity.scala
package de.kasoki.trierbustimetracker

import org.scaloid.common._
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import scala.util.Success
import scala.util.Failure
import scala.collection.mutable.Buffer

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget._
import android.util.Log
import android.content._

import de.kasoki.swtrealtime._

import de.kasoki.trierbustimetracker.adapter.BusTimeAdapter
import de.kasoki.trierbustimetracker.utils.ActionBarHelper
import de.kasoki.trierbustimetracker.utils.AndroidHelper
import de.kasoki.trierbustimetracker.utils.FavoritesManager
import de.kasoki.trierbustimetracker.utils.ShortcutManager

class BusTimeActivity extends SActivity {

    val timesAdapter:BusTimeAdapter = new BusTimeAdapter(this)
    var code:String = ""

    override def onCreate(bundle:Bundle) {
        super.onCreate(bundle)

        invalidateOptionsMenu()

        this.setContentView(R.layout.activity_bustimes)

        val intent = getIntent()

        ActionBarHelper.colorActionBar(this)

        if(intent.hasExtra("FROM_MAIN_ACTIVITY")) {
            ActionBarHelper.enableHomeAsUp(this)
        }

        handleIntent(intent)
    }

    private def handleIntent(intent:Intent) {
        this.code = intent.getStringExtra("BUS_TIME_CODE")

        val listView = list()

        listView.setAdapter(timesAdapter)

        val busStop = BusStop.getBusStopByCode(this.code)

        // set title to bus stop
        this.setTitle(busStop.name)

        // get information
        reload()
    }

    override def onNewIntent(intent:Intent) {
        super.onNewIntent(intent)

        setIntent(intent)
        handleIntent(intent)
    }

    override def onCreateOptionsMenu(menu:Menu):Boolean = {
        // Inflate the menu; this adds items to the action bar if it is present.
        this.getMenuInflater().inflate(R.menu.bustimes_menu, menu)

        val favoritesItem = menu.findItem(R.id.action_favorite)

        val busStop = BusStop.getBusStopByCode(this.code)

        if(FavoritesManager.has(this, busStop)) {
            favoritesItem.setIcon(R.drawable.ic_favorite_filled)
        }

        return true;
    }

    override def onOptionsItemSelected(item:MenuItem):Boolean = {
        item.getItemId() match {
            case android.R.id.home => {
                finish()

                AndroidHelper.slideBack(this)

                return true
            }

            case R.id.action_favorite => {
                val busStop = BusStop.getBusStopByCode(this.code)

                if(FavoritesManager.has(this, busStop)) {
                    item.setIcon(R.drawable.ic_favorite_empty)

                    FavoritesManager.remove(this, busStop)

                    val message = getResources().getString(R.string.favorite_item_deleted, busStop.name)
                    toast(message)
                } else {
                    item.setIcon(R.drawable.ic_favorite_filled)

                    FavoritesManager.add(this, busStop)

                    val message = getResources().getString(R.string.favorite_item_added, busStop.name)
                    toast(message)
                }

                invalidateOptionsMenu()

                return true
            }

            case R.id.action_reload => {
                reload()

                return true
            }

            case R.id.action_add_to_homescreen => {
                val busStop = BusStop.getBusStopByCode(code)

                ShortcutManager.create(this, busStop)

                toast(getString(R.string.shortcut_created, busStop.name))

                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }

    def reload() {
        if(AndroidHelper.isNetworkAvailable(this)) {

            val loadingText = getResources().getString(R.string.loading_dialog_text)

            val progressDialog = spinnerDialog("", loadingText)

            val retryText = getResources().getString(R.string.retry_text)
            val cancelText = getResources().getString(android.R.string.cancel)

            val future = Future {
                val timeout = 5000 // 5s
                BusTime.timeout = timeout
                BusTime.fromStopCode(this.code)
            }

            future onComplete {
                case Success(times) => {

                    runOnUiThread({
                        progressDialog.dismiss()

                        if(times.length > 0) {
                            this.timesAdapter.items = times.sorted

                            this.timesAdapter.notifyDataSetChanged()
                        } else {
                            val message = getResources().getString(R.string.bustime_no_bus)

                            new AlertDialogBuilder("", message) {
                                positiveButton(retryText, reload())
                                setNegativeButton(cancelText, new DialogInterface.OnClickListener {
                                    def onClick(dialog:DialogInterface, which:Int):Unit = closeActivity()
                                })
                                setOnCancelListener(new DialogInterface.OnCancelListener() {
                                    def onCancel(dialog:DialogInterface):Unit = closeActivity()
                                })
                            }.show()
                        }
                    })
                }

                case Failure(t:Throwable) => {
                    progressDialog.dismiss()

                    error(t.toString)

                    val sw = new java.io.StringWriter()
                    t.printStackTrace(new java.io.PrintWriter(sw))
                    val str = sw.toString()

                    error(str)

                    new AlertDialogBuilder("", t.toString) {
                        positiveButton(retryText, reload())
                        setNegativeButton(cancelText, new DialogInterface.OnClickListener {
                            def onClick(dialog:DialogInterface, which:Int):Unit = {
                                if(timesAdapter.items.length > 0) {
                                    dialog.dismiss()
                                } else {
                                    closeActivity()
                                }
                            }
                        })
                        setOnCancelListener(new DialogInterface.OnCancelListener() {
                            def onCancel(dialog:DialogInterface):Unit = {
                                if(timesAdapter.items.length > 0) {
                                    dialog.dismiss()
                                } else {
                                    closeActivity()
                                }
                            }
                        })
                    }.show()
                }
            }

        } else {
            // No connection
            debug("NO NETWORK CONNECTION")

            val noNetwork = getResources().getString(R.string.no_network_connection_text)

            val retryText = getResources().getString(R.string.retry_text)
            val cancelText = getResources().getString(android.R.string.cancel)

            new AlertDialogBuilder("", noNetwork) {
                positiveButton(retryText, reload())
                setNegativeButton(cancelText, new DialogInterface.OnClickListener {
                    def onClick(dialog:DialogInterface, which:Int):Unit = {
                        if(timesAdapter.items.length > 0) {
                            dialog.dismiss()
                        } else {
                            closeActivity()
                        }
                    }
                })
                setOnCancelListener(new DialogInterface.OnCancelListener() {
                    def onCancel(dialog:DialogInterface):Unit = {
                        if(timesAdapter.items.length > 0) {
                            dialog.dismiss()
                        } else {
                            closeActivity()
                        }
                    }
                })
            }.show()
        }
    }

    def closeActivity() {
        finish()

        if(getIntent.hasExtra("FROM_MAIN_ACTIVITY")) {
            AndroidHelper.slideBack(BusTimeActivity.this)
        }
    }

    override def onBackPressed() {
        super.onBackPressed()

        AndroidHelper.slideBack(this)
    }

    def list():ListView = this.findViewById(R.id.busStopListView).asInstanceOf[ListView]
}
